sat/cli/bootprep/image.py

```diff
@@ -370,22 +432,25 @@ def create_images(instance, args):
     for image in input_images:
         image.public_key_id = ims_public_key_id
 
-    validate_unique_image_names(input_images)
-    images_to_create = handle_existing_images(ims_client, input_images, args.overwrite_images,
-                                              args.skip_existing_images, args.dry_run)
+    validate_unique_image_ref_names(input_images)
+
+    # Raises ImageCreateError if validation of IMS bases fails
+    # Do this first to provide information required to render names.
+    validate_image_ims_bases(input_images)
 
     # Raises ImageCreateError if validation of CFS configurations fails
-    validate_image_configurations(images_to_create, cfs_client,
+    validate_image_configurations(input_images, cfs_client,
                                   instance.input_configuration_names, args.dry_run)
 
-    create_verb = ('Creating', 'Would create')[args.dry_run]
-    LOGGER.info(f'{create_verb} {len(images_to_create)} images.')
+    # This is where name properties are first accessed and thus rendered
+    validate_unique_image_names(input_images)
 
-    if not input_images:
-        LOGGER.info('Given input did not define any IMS images')
-        return
+    images_to_create = handle_existing_images(ims_client, input_images, args.overwrite_images,
+                                              args.skip_existing_images, args.dry_run)
+
+    validate_no_overwritten_ims_bases(images_to_create)
 
-    # This can raise ImageCreateError if there are any cycles
+    # This can raise ImageCreateError for unresolvable dependencies
     find_image_dependencies(images_to_create)
```

Comment:
I think this should actually be done before `handle_existing_images`. Suppose you have image A and B in the input file, and B depends on A. Let's say A already exists, and the admin chose to skip creating it. Then `images_to_create` would not contain image A. Then when `images_to_create` is passed to `find_image_dependencies`, B's dependency on A would be unresolvable.

I think it's safe to move this up since it doesn't depend on any names being rendered or anything. It would probably make sense to do right after `validate_unique_image_ref_names`. I'm testing this now.

Reply:
Here's an example of the bug:

```
ncn-m001:~/haasken/sprint-reviews/latest # sat bootprep run --recipe-version latest barebones-customized.yaml
INFO: Validating given input file barebones-customized.yaml
INFO: Input file successfully validated against schema
INFO: Creating 1 CFS configuration(s)
The following CFS configurations already exist: haasken-test-22.07.1. Would you like to skip, overwrite, or abort? [skip,overwrite,abort] skip
INFO: The following CFS configurations already exist and will be skipped: haasken-test-22.07.1
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: Found IMS base for image at index 0: recipe provided by latest version of product csm
INFO: Found IMS base for image at index 1: image from input instance with ref_name="haasken-barebones"
One or more images already exist in IMS with the following names: haasken-cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2. Would you like to skip, overwrite, or abort? [skip,overwrite,abort] skip
INFO: Images with the following names already exist in IMS and will be skipped: haasken-cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2
INFO: Any images which depend on these images will also be skipped.
ERROR: Unable to resolve image from input instance with ref_name="haasken-barebones".
ERROR: Invalid dependencies exist in images to be created. Resolve the dependency issues and try again.
```

Here is after the re-ordering:

```
(c85af60ff17d) sat-container:/sat/share # sat bootprep run --recipe-version latest barebones-customized.ya
ml
INFO: Validating given input file barebones-customized.yaml
INFO: Input file successfully validated against schema
INFO: Creating 1 CFS configuration(s)
The following CFS configurations already exist: haasken-test-22.07.1. Would you like to skip, overwrite, or abort? [skip,overwrite,abort] skip
INFO: The following CFS configurations already exist and will be skipped: haasken-test-22.07.1
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: image at index 1 depends on image at index 0.
INFO: Found IMS base for image at index 0: recipe provided by latest version of product csm
INFO: Found IMS base for image at index 1: image from input instance with ref_name="haasken-barebones"
One or more images already exist in IMS with the following names: haasken-cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2. Would you like to skip, overwrite, or abort? [skip,overwrite,abort] skip
INFO: Images with the following names already exist in IMS and will be skipped: haasken-cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2
INFO: Any images which depend on these images will also be skipped.
INFO: Of the 1 that will be created, 0 have no dependencies and will be created first.
INFO: Creating 1 images.
INFO: Creating images
INFO: Image creation completed successfully
INFO: Creating 1 BOS session template
```